### PR TITLE
Add /state_ids federation API

### PR DESCRIPTION
**Pull request by erikjohnston**
The new API only returns the event_ids for the state, as most requesters will already have the vast majority of the events already.

### CODE REVIEW

File: synapse/federation/federation_client.py [hunk 1]
```diff
@@ -314,9 +314,32 @@ def get_state_for_room(self, destination, room_id, event_id):
...
+
+            pdus = [event_map[e_id] for e_id in state_event_ids]
+            auth_chain = [event_map[e_id] for e_id in auth_event_ids]
+
+            auth_chain.sort(key=lambda e: e.depth)
+
+            defer.returnValue((pdus, auth_chain))
+        except HttpResponseException as e:
+            if e.code == 404:
+                logger.info("Failed to use get_room_state_ids API, falling back")
```
Comment (NegativeMjark) [inline synapse/federation/federation_client.py]: Are you actually falling back here? I'm not seeing any calls to ``transport_layer.get_room_state``

File: synapse/federation/federation_client.py [hunk 2]
```diff
@@ -340,6 +363,64 @@ def get_state_for_room(self, destination, room_id, event_id):
...
+
+        batch_size = 20
+        for i in xrange(0, len(missing_events), batch_size):
+            batch = missing_events[i:i + batch_size]
+
+            deferreds = [
+                self.get_pdu(
+                    destinations=random_server_list(),
+                    event_id=e_id,
+                ).addBoth(lambda r, e: (r, e), e_id)
```
Comment (NegativeMjark) [inline synapse/federation/federation_client.py]: Does this work as a callback for both errors and callbacks?

Comment (erikjohnston) [inline synapse/federation/federation_client.py]: Yes

File: synapse/federation/federation_client.py (hunk 1 above)
Comment (erikjohnston) [inline synapse/federation/federation_client.py]: Woops

File: synapse/federation/federation_client.py [hunk 3]
```diff
@@ -340,6 +374,64 @@ def get_state_for_room(self, destination, room_id, event_id):
...
+                for e_id in batch
+            ]
+
+            res = yield defer.DeferredList(deferreds, consumeErrors=True)
+            for (result, val), (e_id, _) in res:
```
Comment (NegativeMjark) [inline synapse/federation/federation_client.py]: Why is it ``(result,val), (e_id, _)`` here but ``(r, e), e_id)`` in the list comprehension above. What's the difference between ``result`` and ``val``.

**Comment by NegativeMjark** [comment]
LGTM

**Pull request merged**